dns.go

```diff
@@ -82,6 +55,132 @@ func (d *customDialer) archiveDNS(ctx context.Context, address string) (resolved
 	return nil, false, fmt.Errorf("no suitable IP address found for %s", address)
 }
 
+// concurrentDNSLookup tries DNS servers with configurable concurrency
+// - dnsConcurrency <= 1: sequential (one server at a time)
+// - dnsConcurrency > 1: that many servers concurrently
+// - dnsConcurrency == -1: all servers at once (unlimited)
+// Implements early cancellation: stops querying once results are found
+func (d *customDialer) concurrentDNSLookup(ctx context.Context, address string, maxServers int) (ipv4, ipv6 net.IP, errA, errAAAA error) {
+	type result struct {
+		ip         net.IP
+		err        error
+		recordType uint16
+	}
+
+	// Determine effective concurrency
+	concurrency := d.dnsConcurrency
+	if concurrency == -1 {
+		concurrency = maxServers // Unlimited = all servers
+	} else if concurrency <= 0 {
+		concurrency = 1 // Default to sequential
+	}
+
+	// Create cancellable context for early termination
+	workerCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	resultChan := make(chan result, maxServers*2)
+	serverChan := make(chan int, maxServers)
+	var wg sync.WaitGroup
+
+	// Fill server queue with round-robin starting index
+	// Atomically increment and get the starting position
+	startIdx := int(d.dnsRoundRobinIndex.Add(1)-1) % maxServers
+	for i := range maxServers {
+		serverIdx := (startIdx + i) % maxServers
+		serverChan <- serverIdx
+	}
+	close(serverChan)
+
+	// Helper to check if we have all needed results
+	haveAllResults := func() bool {
+		if !d.disableIPv4 && ipv4 == nil {
+			return false
+		}
+		if !d.disableIPv6 && ipv6 == nil {
+			return false
+		}
+		return true
+	}
+
+	// Launch worker goroutines (limited by concurrency)
+	for i := 0; i < concurrency && i < maxServers; i++ {
+		wg.Go(func() {
+			for serverIdx := range serverChan {
+				// Check if context was cancelled before starting queries
+				select {
+				case <-workerCtx.Done():
+					return
+				default:
+				}
+
+				// Query both A and AAAA for this server
+				if !d.disableIPv4 {
+					ip, err := d.lookupIP(workerCtx, address, dns.TypeA, serverIdx)
+					select {
+					case resultChan <- result{ip: ip, err: err, recordType: dns.TypeA}:
+					case <-workerCtx.Done():
+						return
+					}
+				}
+				if !d.disableIPv6 {
+					ip, err := d.lookupIP(workerCtx, address, dns.TypeAAAA, serverIdx)
+					select {
+					case resultChan <- result{ip: ip, err: err, recordType: dns.TypeAAAA}:
+					case <-workerCtx.Done():
+						return
+					}
+				}
+			}
+		})
+	}
+
+	// Close result channel when all workers complete
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+	// Collect results with early termination
+	var ipv4Errors, ipv6Errors []error
+	for res := range resultChan {
+		if res.err == nil {
+			if res.recordType == dns.TypeA && ipv4 == nil {
```

Comment:
While we have the records here, do we have any interest in checking if they are the same?